core/state_processor.go

```diff
@@ -86,11 +91,18 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 // ApplyTransactions returns the generated receipts and vm logs during the
 // execution of the state transition phase.
 func ApplyTransaction(config *ChainConfig, bc *BlockChain, gp *GasPool, statedb *state.StateDB, header *types.Header, tx *types.Transaction, usedGas *big.Int, cfg vm.Config) (*types.Receipt, vm.Logs, *big.Int, error) {
-	_, gas, err := ApplyMessage(NewEnv(statedb, config, bc, tx, header, cfg), tx, gp)
+	env := NewEnv(statedb, config, bc, tx, header, cfg)
+	_, gas, err := ApplyMessage(env, tx, gp)
 	if err != nil {
 		return nil, nil, nil, err
 	}
 
+	for _, codeHash := range env.CodeHashes {
```

Comment:
Minor performance suggestion: consider changing `env.CodeHashes` to `map[common.Hash]struct{}` and `IllegalCodeHashes` to an array and loop over `IllegalCodeHashes` instead, to avoid linear complexity on `env.CodeHashes`

Probably not very significant but most likely `IllegalCodeHashes.length` will be 1 but `env.CodeHashes.length` could be a few thousand for complex txs.